osu.Game/Screens/Edit/EditorBeatmap.cs

```diff
@@ -54,70 +54,71 @@ public BeatmapInfo BeatmapInfo
 
-        public IReadOnlyList<T> HitObjects => beatmap.HitObjects;
+        public IReadOnlyList<HitObject> HitObjects => beatmap.HitObjects;
 
         IReadOnlyList<HitObject> IBeatmap.HitObjects => beatmap.HitObjects;
 
         public IEnumerable<BeatmapStatistic> GetStatistics() => beatmap.GetStatistics();
 
-        public IBeatmap Clone() => (EditorBeatmap<T>)MemberwiseClone();
+        public IBeatmap Clone() => (EditorBeatmap)MemberwiseClone();
 
         /// <summary>
-        /// Adds a <see cref="HitObject"/> to this <see cref="EditorBeatmap{T}"/>.
+        /// Adds a <see cref="HitObject"/> to this <see cref="EditorBeatmap"/>.
         /// </summary>
         /// <param name="hitObject">The <see cref="HitObject"/> to add.</param>
-        public void Add(T hitObject)
+        public void Add(HitObject hitObject)
         {
             trackStartTime(hitObject);
 
             // Preserve existing sorting order in the beatmap
-            var insertionIndex = beatmap.HitObjects.FindLastIndex(h => h.StartTime <= hitObject.StartTime);
-            beatmap.HitObjects.Insert(insertionIndex + 1, hitObject);
+            var insertionIndex = findInsertionIndex(beatmap.HitObjects, hitObject.StartTime);
+            ((IList)beatmap.HitObjects).Insert(insertionIndex + 1, hitObject);
 
             HitObjectAdded?.Invoke(hitObject);
         }
 
         /// <summary>
-        /// Removes a <see cref="HitObject"/> from this <see cref="EditorBeatmap{T}"/>.
+        /// Removes a <see cref="HitObject"/> from this <see cref="EditorBeatmap"/>.
         /// </summary>
         /// <param name="hitObject">The <see cref="HitObject"/> to add.</param>
-        public void Remove(T hitObject)
+        public void Remove(HitObject hitObject)
         {
-            if (beatmap.HitObjects.Remove(hitObject))
-            {
-                var bindable = startTimeBindables[hitObject];
-                bindable.UnbindAll();
+            if (!((IList)beatmap.HitObjects).Contains(hitObject))
+                return;
 
-                startTimeBindables.Remove(hitObject);
-                HitObjectRemoved?.Invoke(hitObject);
-            }
+            ((IList)beatmap.HitObjects).Remove(hitObject);
+
+            var bindable = startTimeBindables[hitObject];
+            bindable.UnbindAll();
+
+            startTimeBindables.Remove(hitObject);
+            HitObjectRemoved?.Invoke(hitObject);
         }
 
-        private void trackStartTime(T hitObject)
+        private void trackStartTime(HitObject hitObject)
         {
             startTimeBindables[hitObject] = hitObject.StartTimeBindable.GetBoundCopy();
             startTimeBindables[hitObject].ValueChanged += _ =>
             {
                 // For now we'll remove and re-add the hitobject. This is not optimal and can be improved if required.
-                beatmap.HitObjects.Remove(hitObject);
+                ((IList)beatmap.HitObjects).Remove(hitObject);
```

Comment:
looked at this PR a few times and still not really sure about this solution... but let's go with it for now, since it's quite localised.

Reply:
I'm willing to accept any alternative. One possible solution is to use `List<HitObject>`, but you cannot have co-variant types with non-interfaces so there can't also be a `List<THitObject>`.